pr_agent/servers/bitbucket_server_webhook.py

```diff
@@ -13,35 +14,55 @@
 from pr_agent.agent.pr_agent import PRAgent
 from pr_agent.config_loader import get_settings
 from pr_agent.log import get_logger
+from pr_agent.servers.utils import verify_signature
 
 router = APIRouter()
 
 
-def handle_request(background_tasks: BackgroundTasks, url: str, body: str, log_context: dict):
+def handle_request(
+    background_tasks: BackgroundTasks, url: str, body: str, log_context: dict
+):
     log_context["action"] = body
-    log_context["event"] = "pull_request" if body == "review" else "comment"
     log_context["api_url"] = url
     with get_logger().contextualize(**log_context):
         background_tasks.add_task(PRAgent().handle_request, url, body)
 
 
-@router.post("/webhook")
+@router.post("/")
 async def handle_webhook(background_tasks: BackgroundTasks, request: Request):
     log_context = {"server_type": "bitbucket_server"}
     data = await request.json()
     get_logger().info(json.dumps(data))
 
-    pr_id = data['pullRequest']['id']
-    repository_name = data['pullRequest']['toRef']['repository']['slug']
-    project_name = data['pullRequest']['toRef']['repository']['project']['key']
+    webhook_secret = getattr(get_settings().bitbucket_server, "webhook_secret", None)
+    if webhook_secret:
+        body_bytes = await request.body()
+        signature_header = request.headers.get("x-hub-signature", None)
+        verify_signature(body_bytes, webhook_secret, signature_header)
+
+    pr_id = data["pullRequest"]["id"]
+    repository_name = data["pullRequest"]["toRef"]["repository"]["slug"]
+    project_name = data["pullRequest"]["toRef"]["repository"]["project"]["key"]
     bitbucket_server = get_settings().get("BITBUCKET_SERVER.URL")
     pr_url = f"{bitbucket_server}/projects/{project_name}/repos/{repository_name}/pull-requests/{pr_id}"
 
     log_context["api_url"] = pr_url
     log_context["event"] = "pull_request"
 
-    handle_request(background_tasks, pr_url, "review", log_context)
-    return JSONResponse(status_code=status.HTTP_200_OK, content=jsonable_encoder({"message": "success"}))
+    if data["eventKey"] == "pr:opened":
+        body = "review"
+    elif data["eventKey"] == "pr:comment:added":
+        body = data["comment"]["text"]
+    else:
+        return JSONResponse(
```

Comment:
I think we should return 200 in this case, this is not an error, unless you know differently?

Reply:
Well we can, but you have to configure which events you want to be pushed as webhook. If you push some events that are not supported it could be a bad request, or could be a 200 and who cares or, could also be a 501 not implemented which makes also sense.

Reply:
I'm asking before in some cases, after a certain threshold of errors the webhook is disabled.